### PR TITLE
Adds test merged PR title and author to the revision data

### LINKED ISSUE #23550: Duplication in list of test merges
image

@MrStonedOne

### PR DESCRIPTION
**Pull request by Cyberboss**
:cl: Cyberboss
add: Test merged PRs are now more detailed
/:cl:

Revdata.testmerge is now an assoc list with values that are parsed json that looks like this: https://api.github.com/repositories/3234987/pulls/23359

### CODE REVIEW

**Comment by coiax** [comment]
WE SHOULD TEST MERGE THIS

**Comment by Cyberboss** [comment]
Muh recursions...

On Jan 26, 2017 1:10 PM, "coiax" <notifications@github.com> wrote:

WE SHOULD TEST MERGE THIS

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/tgstation/tgstation/pull/23359#issuecomment-275464569>,
or mute the thread

**Comment by Cyberboss** [comment]
DNM bugs

File: code/datums/helper_datums/getrev.dm [hunk 1]
```diff
@@ -21,13 +23,58 @@ var/global/datum/getrev/revdata = new()
...
+			     )   _ /  / _)
+			 /\  )/\/ ||  | )_)
+			<  >      |(,,) )__)
+			 ||      /    \)___)\
+			 | \____(      )___) )___
+			  \______(_______;;; __;;;
+		*/
+		var/url = "https://api.github.com/repositories/[config.githubrepoid]/pulls/[line].json"
+		var/temp_file = "prcheck[line].json"
+		var/command = "powershell -Command \"curl -O [temp_file] [url]\""
```
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: This is unlikely to be something i'll allow. As it is, right now you've given anybody with vv rights the ability to run remote code on the server, and i don't want to be playing games trying to do shell escaping in byond if we can avoid it.

There MUST be some way to get this info natively in byond.

If you believe you can not, I want a detailed explanation as to why you have to run cmd to run powershell to run curl to get a simple web page.

Comment (coiax) [inline code/datums/helper_datums/getrev.dm]: Isn't test merging some sort of script you run on the server? Can't we have that script populate the place where it puts the testmerge information with the title and author as well?

Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: I'm not even entertaining that idea until i fully understand why `world`.`Export()` won't work.

**Comment by Cyberboss** [comment]
@MrStonedOne World.Export won't work because the GitHub API requires HTTPS
which, of course, BYOND doesn't support.

**Comment by RemieRichards** [comment]
@Cyberboss so why not:
```Byond -> Something that can HTTPS -> Github -> Something that can HTTPS ->Byond```?

**Comment by MrStonedOne** [comment]
remie, the word you are looking for is http proxy.

Also, i strongly remember there being a way to tell github to not redirect the https connection.

but first, can you actually confirm you tried the http version in `world`.`Export()`.

**Comment by RemieRichards** [comment]
@MrStonedOne who cares about the name I still got the concept right it's midnight and I have to go to bed gosh.

**Comment by Cyberboss** [comment]
Yep it's github enforcing the HTTPS

**Comment by MrStonedOne** [comment]
What did you get when you tried it?

**Comment by MrStonedOne** [comment]
also, what's up with that powershell bit?

**Comment by Cyberboss** [comment]
Http attempts just return null. The powershell is a workaround to just curl the page and save it to a file that we then read.

**Comment by MrStonedOne** [comment]
but... why is powershell needed?

**Comment by Cyberboss** [comment]
Cause basically CMD is so bad it doesn't have basic meet tools

**Comment by MrStonedOne** [comment]
oh wait. curl is included in ps?

**Comment by Cyberboss** [comment]
Yep

File: code/datums/helper_datums/getrev.dm [hunk 2]
```diff
@@ -21,13 +23,58 @@ var/global/datum/getrev/revdata = new()
...
+			     ) /^\) ^\/ _)
+			     )   /^\/   _)
+			     )   _ /  / _)
+			 /\  )/\/ ||  | )_)
+			<  >      |(,,) )__)
+			 ||      /    \)___)\
+			 | \____(      )___) )___
+			  \______(_______;;; __;;;
+		*/
+		var/url = "https://api.github.com/repositories/[config.githubrepoid]/pulls/[line].json"
```
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: You need to validate that both config.githubrepoid and line are numbers (and/or text of numbers) or trialmins, AdminCanidates, etc could abuse this to take control of the server and make it execute raw code.

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]

File: code/datums/helper_datums/getrev.dm [hunk 3]
```diff
@@ -21,13 +23,63 @@ var/global/datum/getrev/revdata = new()
...
-				world.log << "Test merge active of PR #[line]"
+			world.log << "Test merge active of PR #[line]"
 		world.log << "Based off master commit [parentcommit]"
 	else
 		world.log << parentcommit
 	world.log << "Current map - [MAP_NAME]" //can't think of anywhere better to put it
 
+/datum/getrev/proc/DownloadPRDetails()
+	if(!text2num(config.githubrepoid))
+		world.log << "Invalid github repo id: [config.githubrepoid]"
```
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: What if somebody wants to just disable this feature

File: code/controllers/configuration.dm [hunk 4]
```diff
@@ -69,6 +69,7 @@
 	var/forumurl = "http://tgstation13.org/phpBB/index.php" //default forums
 	var/rulesurl = "http://www.tgstation13.org/wiki/Rules" // default rules
 	var/githuburl = "https://www.github.com/tgstation/-tg-station" //default github
+	var/githubrepoid = "3234987"  //default github
```
Comment (MrStonedOne) [inline code/controllers/configuration.dm]: Don't have a default in here, have it in the config. that way somebody can disable the feature by commenting this out.

Giving server ops the ability to control if the game decides to access third party apis or run shell() commands is important enough, and making them have to define the config as a blank string is wasteful.

File: code/controllers/configuration.dm [hunk 5]
```diff
@@ -360,6 +361,8 @@
 					config.rulesurl = value
 				if("githuburl")
 					config.githuburl = value
+				if("githubrepoid")
+					config.githuburl = value
```
Comment (MrStonedOne) [inline code/controllers/configuration.dm]: convert to a number here, then check isnum below, its cheaper and how most things do it.

File: code/datums/helper_datums/getrev.dm (hunk 3 above)
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: Make this a isnum() (like above) but add a if (!config.githubrepoid) above this if with just a simple return.

File: code/datums/helper_datums/getrev.dm [hunk 6]
```diff
@@ -21,13 +23,63 @@ var/global/datum/getrev/revdata = new()
...
+
+/datum/getrev/proc/GetTestMergeInfo()
+	if(!testmerge.len)
+		return ""
+	. = "The following pull requests are currently test merged:<br/>"
+	for(var/line in testmerge)
+		var/details = ""
+		if(has_pr_details)
+			details = ": '" + testmerge[line]["title"] + "' by " + testmerge[line]["user"]["login"]
+		. += "<a href='[config.githuburl]/pull/[line]'>#[line][details]</a><br/>"
```
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: Needs html sanitation or somebody could inject html into the chatbox by putting it in the title.

File: code/controllers/configuration.dm (hunk 5 above)
Comment (Cyberboss) [inline code/controllers/configuration.dm]: We can't convert it to a num here because then it becomes an exponential and the url won't work properly

Comment (MrStonedOne) [inline code/controllers/configuration.dm]: http://www.byond.com/docs/ref/info.html#/proc/num2text

File: code/datums/helper_datums/getrev.dm (hunk 6 above)
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: http://www.byond.com/docs/ref/info.html#/proc/html_encode

**Comment by optimumtact** [comment]
what is the state of this

**Comment by MrStonedOne** [comment]
I'll test merge it to see.

**Comment by MrStonedOne** [comment]
```
The following pull requests are currently test merged:
>#23493 
>#23494 
>#23359 
># 
>

```

**Comment by MrStonedOne** [comment]
```
The following pull requests are currently test merged:<br/><a href='3234987/pull/23493'>#23493</a><br/><a href='3234987/pull/23494'>#23494</a><br/><a href='3234987/pull/23359'>#23359</a><br/><a href='3234987/pull/'>#</a><br/></div><br>
```

File: code/controllers/configuration.dm (hunk 5 above)
Comment (MrStonedOne) [inline code/controllers/configuration.dm]: You kept the varname the same when you copy and pasted this!

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]

File: code/datums/helper_datums/getrev.dm [hunk 7]
```diff
@@ -21,13 +23,65 @@ var/global/datum/getrev/revdata = new()
...
+			 ||      /    \)___)\
+			 | \____(      )___) )___
+			  \______(_______;;; __;;;
+			*/
+		if(!isnum(line))
+			world.log << "Invalid PR number: [line]"
+			return
+		var/url = "https://api.github.com/repositories/[config.githubrepoid]/pulls/[line].json"
+		var/temp_file = "prcheck[line].json"
+		var/command = "powershell -Command \"curl -O [temp_file] [url]\""
```
Comment (AnturK) [inline code/datums/helper_datums/getrev.dm]: 🐧 🔫

**Comment by coiax** [comment]
image

PLEASE FINISH THIS PR

**Comment by MrStonedOne** [comment]
Test merged on both servers

File: code/datums/helper_datums/getrev.dm (hunk 7 above)
Comment (Iamgoofball) [inline code/datums/helper_datums/getrev.dm]: :+1:

**Comment by AnturK** [comment]
Any issues left with this ?

**Comment by MrStonedOne** [comment]
most likely, every time it's been test merged it's done nothing.

File: code/datums/helper_datums/getrev.dm (hunk 7 above)
Comment (JJRcop) [inline code/datums/helper_datums/getrev.dm]: https://github.com/PowerShell/powershell/ ⚡️🐚 ➡️ 🐧

**Comment by Cyberboss** [comment]
Still not working, but just realized I may be making an assumption about the contents of prtest.lk. Mind sharing one?

File: code/datums/helper_datums/getrev.dm [hunk 8]
```diff
@@ -21,13 +25,70 @@ var/global/datum/getrev/revdata = new()
...
+		if(!testmerge[line])
+			world.log << "PR details download failed: null details returned"
+			return
+		CHECK_TICK
+	has_pr_details = TRUE
+
+/datum/getrev/proc/GetTestMergeInfo()
+	if(!testmerge.len)
+		return ""
+	. = "The following pull requests are currently test merged:<br><br/>"
```
Comment (MrStonedOne) [inline code/datums/helper_datums/getrev.dm]: stop with the `/>`

Byond doesn't recognize it and the chatbox is not a real browse window, its byond fake html window.

**Comment by coiax** [comment]
WHEN IS THIS GOING TO GET MERGED

**Comment by Cyberboss** [comment]
WHEN IT WORKS

**Comment by Cyberboss** [comment]
```
PR details download failed: shell exited with code: 1
```

**Comment by Cyberboss** [comment]
This is the wrong approach and hacky. It needs to be something done from the testmerge tools. It's also holding up a fix so I'm gonna close this.

**Comment by Cyberboss** [comment]
Gonna try something else

File: code/__HELPERS/unsorted.dm [hunk 9]
```diff
@@ -1336,3 +1336,49 @@ proc/pick_closest_path(value, list/matches = get_fancy_list_of_atom_types())
 			if(W.ini_dir == dir_to_check || W.ini_dir == FULLTILE_WINDOW_DIR || dir_to_check == FULLTILE_WINDOW_DIR)
 				return FALSE
 	return TRUE
+
+//WHATEVER YOU USE THIS FOR MUST BE SANITIZED TO SHIT, IT USES SHELL
+//It also sleeps
+
+//Set this to TRUE before calling
+//This prevents RCEs from badmins
+//kevinz000 if you touch this I will hunt you down
```
Comment (PKPenguin321) [inline code/__HELPERS/unsorted.dm]: :ok_hand:

**Comment by Cyberboss** [comment]
It finally fucking works! This can be merged with MSO's permission.

**Comment by optimumtact** [comment]
it's conflicting, but can @MrStonedOne approve this so I can finally merge it

**Review by MrsTonedOne — APPROVED** [review]
I see no glaring security issues.

**Comment by JJRcop** [comment]
Hey did you know that powershell is actually cross-platform now? https://github.com/PowerShell/PowerShell

**Pull request merged**